projects/RabbitMQ.Client/client/TaskExtensions.cs

```diff
@@ -0,0 +1,96 @@
+﻿// This source code is dual-licensed under the Apache License, version
+// 2.0, and the Mozilla Public License, version 2.0.
+//
+// The APL v2.0:
+//
+//---------------------------------------------------------------------------
+//   Copyright (c) 2007-2020 VMware, Inc.
+//
+//   Licensed under the Apache License, Version 2.0 (the "License");
+//   you may not use this file except in compliance with the License.
+//   You may obtain a copy of the License at
+//
+//       https://www.apache.org/licenses/LICENSE-2.0
+//
+//   Unless required by applicable law or agreed to in writing, software
+//   distributed under the License is distributed on an "AS IS" BASIS,
+//   WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+//   See the License for the specific language governing permissions and
+//   limitations under the License.
+//---------------------------------------------------------------------------
+//
+// The MPL v2.0:
+//
+//---------------------------------------------------------------------------
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+//
+//  Copyright (c) 2007-2020 VMware, Inc.  All rights reserved.
+//---------------------------------------------------------------------------
+
+using System;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace RabbitMQ.Client
+{
+    internal static class TaskExtensions
+    {
+#if !NET6_0_OR_GREATER
+        private static readonly TaskContinuationOptions s_tco = TaskContinuationOptions.OnlyOnFaulted | TaskContinuationOptions.ExecuteSynchronously;
+#endif
+
+        private static void continuation(Task t, object s) => t.Exception.Handle(e => true);
+
+        public static async Task TimeoutAfter(this Task task, TimeSpan timeout)
+        {
+#if NET6_0_OR_GREATER
+            await task.WaitAsync(timeout)
+                .ConfigureAwait(false);
+#else
+            if (task == await Task.WhenAny(task, Task.Delay(timeout)).ConfigureAwait(false))
```

Comment:
This doesn't cancel the timer task in the cases when the first task successfully completed which will lead to loads to tasks hanging around and probably issues with the timer queue.

See also https://github.com/davidfowl/AspNetCoreDiagnosticScenarios/blob/master/AsyncGuidance.md#using-a-timeout

Reply:
Why not just copy from the framework?

https://source.dot.net/#System.Private.CoreLib/src/libraries/System.Private.CoreLib/src/System/Threading/Tasks/Task.cs,2854

Reply:
And consider naming it `WaitAsync` and drop the implementation post .NET 6.0.

Reply:
And add a `CancellationToken` parameter. Most, if not all asynchronous methods should have one.

Reply:
Yep. See https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1420

Reply:
https://github.com/rabbitmq/rabbitmq-dotnet-client/issues/1425

Reply:
PS I'd really rather not copy framework code but I'll look into that option.

Reply:
> PS I'd really rather not copy framework code but I'll look into that option.

If it's source compatible, beats writing my own.